### DISCUSSION TITLE
icon-text-fit-padding interpolate expressions

### DISCUSSION THREAD

**Discussion by BlancHeart** [maplibre/maplibre-gl-js#1985] [Q&A] [answered]
Hello. I'm a relatively novice user of maplibre and typescript.

I tried to apply interpolate expression based on zoom to [icon-text-fit-padding](https://maplibre.org/maplibre-gl-js-docs/style-spec/layers/#layout-symbol-icon-text-fit-padding)
There's not many documentation, examples, or google results, so I asked ChatGPT.

`'icon-text-fit-padding': [
      'interpolate',
      ['linear'],
      ['zoom'],
      0,
      4,
      22,
      8
    ]`
The first attempt didn't work, throw error:
`icon-text-fit-padding[4]: Expected array<number, 4> but found number instead.`

OK, maybe it make sense to have arrays with four element, like the default format [0,0,0,0].
ChatGPT's 2nd attempt:
`'icon-text-fit-padding': [  'interpolate',  ['linear'],
  ['zoom'],
  0,
  [4, 8, 12, 16],
  22,
  [8, 16, 24, 32]
]`
However, it throws:
`icon-text-fit-padding[4][0]: Expression name must be a string, but found number instead. If you wanted a literal array, use ["literal", [...]].`
I had no idea but to ask the Chat again.

So the 3rd generation is:
`'icon-text-fit-padding': [  'interpolate',  ['linear'],
  ['zoom'],
  0,
  ['literal', [4, 8, 12, 16]],
  22,
  ['literal', [8, 16, 24, 32]]
]`
It seems working in the runtime but typesciprt is not happy about it.
`Type at positions 3 through 6 in source is not compatible with type at position 3 in target.`

I checked through ExpressionSpecification in type definition, and looks like the 2nd attempt should make the most sense.
Does anyone know the correct expression format?

**Comment by HarelM** [comment, marked as answer]
If it's working in runtime, it's the correct solution.
We have typescript issues when it comes to the expressions, unfortunately...

**Reply by BlancHeart** [reply]
Thanks. 
After reading more documentation, I think literal expression making sense now.
from [expression type system](https://maplibre.org/maplibre-gl-js-docs/style-spec/expressions/#type-system)

> If an expression accepts an array argument and the user supplies an array literal, that array must be wrapped in a literal expression (see the examples below). When GL-JS encounters an array in a style-spec property value, it will assume that the array is an expression and try to parse it; the library has no way to distinguish between an expression which failed validation and an array literal unless the developer makes this distinction explicit with the literal operator.

**Comment by wipfli** [comment]
You made a map with ChatGPT, @BlancHeart?

**Reply by BlancHeart** [reply]
Not really.
I usually rely on those examples to learn, then read API doc. However, documenation, such as this case, is not clear sometimes, and I can't find any meaningful answer through Google or StackOverflow.
Since we have ChatGPT now, it's much faster to get a starting point when it can link several different knowledge together, though many times is not accurate, and need feed the answer back to correct it.
